library/src/main/java/com/getstream/sdk/chat/adapter/MessageListItem.java

```diff
@@ -40,11 +49,103 @@ public MessageListItem(List<User> users) {
         this.messageReadBy = new ArrayList<>();
     }
 
+    public MessageListItem copy() {
+        MessageListItem clone = new MessageListItem(message, positions, messageMine);
+        clone.date = date;
+        clone.type = type;
+        clone.users = users;
+        clone.messageReadBy.addAll(messageReadBy);
+        return clone;
+    }
+
+    boolean samePositions(List<MessageViewHolderFactory.Position> a, List<MessageViewHolderFactory.Position> b) {
+        if ((a == null && b != null) || (a != null && b == null)) {
+            return false;
+        }
+        if (a == null) {
+            return true;
+        }
+        if (a.size() != b.size()) {
+            return false;
+        }
+        for (int i = 0; i < a.size(); i++) {
+            if (!a.get(i).equals(b.get(i))) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    boolean sameReads(List<ChannelUserRead>a, List<ChannelUserRead> b){
+        if ((a == null && b != null) || (a != null && b == null)) {
+            return false;
+        }
+        if (a == null) {
+            return true;
+        }
+        if (a.size() == 0 && b.size() == 0) {
+            return true;
+        }
+        if (a.size() != b.size()) {
+            return false;
+        }
+        for (int i = 0; i < a.size(); i++) {
+            if (!a.get(i).equals(b.get(i))) {
+                return false;
+            }
+        }
+        return true;
```

Comment:
this can be simplified by:
1. not allowing null for these lists (you can change it int he message list item constructor)
2. (didn't try this) but i think there are some relevant util functions: https://developer.android.com/reference/java/util/Arrays

Reply:
yeah let's do that once we have some more tests in place; atm this is ugly but (I think) working ok